src/more_funcs/md4.c

```diff
@@ -0,0 +1,216 @@
+
+    /* Round 3 */
+
+    ROUND3(A,B,C,D,0,3);
+    ROUND3(D,A,B,C,8,9);
+    ROUND3(C,D,A,B,4,11);
+    ROUND3(B,C,D,A,12,15);
+
+    ROUND3(A,B,C,D,2,3);
+    ROUND3(D,A,B,C,10,9);
+    ROUND3(C,D,A,B,6,11);
+    ROUND3(B,C,D,A,14,15);
+
+    ROUND3(A,B,C,D,1,3);
+    ROUND3(D,A,B,C,9,9);
+    ROUND3(C,D,A,B,5,11);
+    ROUND3(B,C,D,A,13,15);
+
+    ROUND3(A,B,C,D,3,3);
+    ROUND3(D,A,B,C,11,9);
+    ROUND3(C,D,A,B,7,11);
+    ROUND3(B,C,D,A,15,15);
+
+    ctx->a += A;
+    ctx->b += B;
+    ctx->c += C;
+    ctx->d += D;
+}
+
+void MD4_Update(MD4Ctx* ctx, u32* input, u32 byteCount) {
+    s32 unused[2];
+    u32 tempOffset;
+    u8* currentBlock;
+    u32 tempByteCount;
+    u8 buffer[0x40];
+    u32 i;
+
+    // If no bits are provided or the hash is already finalized, return early
+    if (byteCount == 0 && ctx->finalized != FALSE) {
+        return;
+    }
+
+    if (ctx->finalized != FALSE) {
+        return;
+    }
+
+    tempByteCount = byteCount;
+    for (i = 0, currentBlock = ctx->buffer; tempByteCount != 0; i++, currentBlock++) {
+        tempByteCount += *currentBlock;
+        *currentBlock = tempByteCount;
+        tempByteCount >>= 8;
+    }
+
+    // Process input if it's a complete block
+    if (byteCount == MD4_BLOCK_LENGTH) {
+        MD4_Transform(ctx, input);
+        return;
+    }
+
+    // If input length is more than a full block, return early
+    if (byteCount > MD4_BLOCK_LENGTH) {
+        return;
+    }
+
+    // If input length is less than a full block
+    tempOffset = byteCount >> 3;
+    for (i = 0; i <= tempOffset; i++) {
+        buffer[i] = ((u8*) input)[i];
+    }
+
+    // Since the input is less than a full block, we will prepare our own buffer
+    for (i = tempOffset + 1; i < ARRAY_COUNT(buffer); i++) buffer[i] = 0;
+
+    buffer[tempOffset] = (buffer[tempOffset] | (1 << (7 - (byteCount & 7)))) & ~((1 << (7 - (byteCount & 7))) - 1);
+
+    if (tempOffset < 0x38) {
+        for (i = 0; i < 8; i++) {
+            buffer[0x38 + i] = ctx->buffer[i];
+        }
+
+        MD4_Transform(ctx, (u32*) buffer);
+    } else {
+        MD4_Transform(ctx, (u32*) buffer);
+
+        for (i = 0; i < 0x38; i++) {
+            buffer[i] = 0;
+        }
+
+        for (i = 0; i < 8; i++) {
+            buffer[0x38 + i] = ctx->buffer[i];
+        }
+        MD4_Transform(ctx, (u32*) buffer);
+    }
+    ctx->finalized = TRUE;
+}
+
+MD4Ctx* MD4_Finalize(u32* input, s32 length) {
+    s32 tempBytes;
+    u32 remainingBytes;
+    u32* currentBlock;
+
+    MD4_Init(&MD4_GlobalContext);
+    remainingBytes = 0;
+    if (length >= 0x40) {
+        currentBlock = input;
+        tempBytes = 0x40;
+        do {
+            MD4_Update(&MD4_GlobalContext, currentBlock, 0x200);
+            remainingBytes = tempBytes;
+            tempBytes += 0x40;
+            currentBlock += 0x10;
+        } while (length >= tempBytes);
+        // required to match
+        if (!tempBytes && !tempBytes && !tempBytes) {
```

Comment:
Do you need all 3 of these, and does it need to be not-ed? It usually can just be if(sym) in my experience

Reply:
I tried to remove this, format it differently, etc. I removed the **!** but was unable to find anything better that matched. Apologies.

Reply:
no apology needed - thanks for trying!